### PR TITLE
#1724 - E2E/Units Tests - Review Queue Consumers E2E Tests

### PR DESCRIPTION
**Pull request by andrewsignori-aot**
- Created the queue-consumers helper method `createTestingAppModule` to start the queue-consumers module for all E2E tests.
- Created the helper method `describeProcessorRootTest` with the intention to create a similar one for the scheduler in the future.
- Extended the mocked Zeebe module.
- Added the mock helper `@golevelup/ts-jest` referenced on Nestjs documentation (https://docs.nestjs.com/fundamentals/testing). The idea is to validate how useful it can be and assess if it is needed or not.
- New scenarios added for `start-application-assessment` and `cancel-application-assessment`, as shown below.
- `ecert-full-time-process-integration.scheduler.e2e-spec` is not the focus of this PR. It was just adjusted to fit better in the new E2E structure.

Processor E2E tests for start-application-assessment
    √ Should throw an error when the workflow createProcessInstance method throws an error.
    √ Should invoke the workflow create instance method with the received job parameters.

Processor E2E tests for cancel-application-assessment
    √ Should cancel pending disbursements and rollback overawards when the cancelled application has overawards and also one sent and one pending disbursements.
    √ Should cancel workflow and log a warning when the workflowInstanceId is not present.
    √ Should throw an error and call job.discard when the application is not in the expected status.
    √ Should throw an error and call job.discard when the assessment id was not found.

### CODE REVIEW

File: sources/packages/backend/apps/queue-consumers/src/processors/assessment/cancel-application-assessment.processor.ts [hunk 1]
```diff
@@ -44,6 +44,7 @@ export class CancelApplicationAssessmentProcessor {
     summary.info(
       `Cancelling application assessment id ${job.data.assessmentId}`,
     );
+
```
Comment (ann-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/assessment/cancel-application-assessment.processor.ts:47]: looks like this file has only this change

File: sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+    it("Should cancel the assessment pending disbursements and rollback overawards when the cancelled application has overawards and also one sent and one pending disbursements.", async () => {
+      // Arrange
+      const workflowInstanceId = "2251799814028835";
+      // Application and disbursements.
+      const application = await saveFakeApplicationDisbursements(
+        appDataSource,
+        null,
+        {
+          applicationStatus: ApplicationStatus.Cancelled,
```
Comment (ann-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts:63]: maybe we can pass `complete` status here and overwrite the status to `cancelled` later in the test, so that, the below portions gets covered by the function `saveFakeApplicationDisbursements`, 
image thoughts?

Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts:63]: I am not sure if I am following. Please keep in mind that COE status is not being tested here, only `disbursementScheduleStatus`. Please let me know what I am missing 😉

File: sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+          createSecondDisbursement: true,
+        },
+      );
+      // Adjust assessment
```
[the remark this answers is not in the document]
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts]: Added periods.

File: sources/packages/backend/apps/api/src/testHelpers/testing-modules/testing-modules-helper.ts [hunk 4]
```diff
@@ -23,7 +23,7 @@ export class CreateTestingModuleResult {
 export async function createTestingAppModule(): Promise<CreateTestingModuleResult> {
   await KeycloakConfig.load();
   const module: TestingModule = await Test.createTestingModule({
-    imports: [AppModule, createMockedZeebeModule(), MockedQueueModule],
+    imports: [AppModule, createZeebeModuleMock(), MockedQueueModule],
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/testHelpers/testing-modules/testing-modules-helper.ts:26]: Just checking. Any particular reason for changing the name to `createZeebeModuleMock` ?

File: sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+      );
+      expect(result.summary).toContain(
+        "Workflow instance successfully cancelled.",
+      );
+      expect(result.summary).toContain("Assessment cancelled with success.");
+
+      // Assert that overawards were soft deleted.
+      const updatedOveraward = await disbursementOverawardRepo.findOne({
+        where: { id: overaward.id },
+        withDeleted: true,
```
Comment (ann-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts:103]: 👍

File: sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/start-application-assessment.processor.e2e-spec.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    it("Should throw an error when the workflow createProcessInstance method throws an error.", async () => {
+      // Arrange
+      const dummyException = new Error("Dummy error");
+      const job = createMock<Job<StartAssessmentQueueInDTO>>();
+      zbClientMock.createProcessInstance = jest.fn().mockImplementation(() => {
+        throw dummyException;
+      });
+
+      // Act and Assert.
+      await expect(processor.startAssessment(job)).rejects.toBe(dummyException);
```
Comment (ann-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/start-application-assessment.processor.e2e-spec.ts:41]: just to double confirm, this is testing line 36 right 
image

Comment (ann-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/start-application-assessment.processor.e2e-spec.ts:41]: yes, it is. Thanks

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/_tests_/ecert-full-time-process-integration.scheduler.e2e-spec.ts [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+    });
+
+    it("Should execute overawards deductions and calculate awards effective value", async () => {
+      // Arrange
+
+      // User
+      const savedUser = await userRepo.save(createFakeUser());
+      // Student.
+      const savedStudent = await studentRepo.save(createFakeStudent(savedUser));
+      // Student SIN Validation
```
Comment (ann-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/_tests_/ecert-full-time-process-integration.scheduler.e2e-spec.ts:84]: period and line 197

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/_tests_/ecert-full-time-process-integration.scheduler.e2e-spec.ts [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+      firstSchedule.coeStatus = COEStatus.completed;
+      firstSchedule.disbursementScheduleStatus =
+        DisbursementScheduleStatus.Pending;
+      firstSchedule.msfaaNumber = savedMSFAANumber;
+      fakeOriginalAssessment.disbursementSchedules = [firstSchedule];
+      // Offering.
+      const fakeOffering = createFakeEducationProgramOffering({
+        auditUser: savedUser,
+      });
+      fakeOffering.offeringIntensity = OfferingIntensity.fullTime;
```
Comment (ann-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/_tests_/ecert-full-time-process-integration.scheduler.e2e-spec.ts:147]: its by default `fullTime`

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/_tests_/ecert-full-time-process-integration.scheduler.e2e-spec.ts [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+      const job = createMock<Job<void>>();
+
+      // Act
+      const result = await processor.processFullTimeECert(job);
+
+      // Assert
+
+      // Assert Canada Loan overawards were deducted.
+      const hasCanadaLoanOverawardDeduction =
+        await disbursementOverawardRepo.exist({
```
Comment (ann-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/_tests_/ecert-full-time-process-integration.scheduler.e2e-spec.ts:175]: nice

File: sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts (hunk 2 above)
Comment (ann-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts:63]: ya, I know COE status is not tested here, it's my opinion, not a blocker.

Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts:63]: @ann-aot maybe I am still missing the point here. I did not get exactly what your option was. Happy to talk if needed.

File: sources/packages/backend/apps/api/src/testHelpers/testing-modules/testing-modules-helper.ts (hunk 4 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/testHelpers/testing-modules/testing-modules-helper.ts:26]: The class was moved to the lib and it seems more aligned with others.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/tests/ecert-full-time-process-integration.scheduler.e2e-spec.ts
Comment (ann-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/tests/ecert-full-time-process-integration.scheduler.e2e-spec.ts:1]: why is this deleted?

Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/tests/ecert-full-time-process-integration.scheduler.e2e-spec.ts:1]: It was not deleted it was moved. It was in the folder "tests" and we are now using a folder "_tests_".
All my attempts to force git to understand that it was moved failed.
I tried even with "git mv", but in the end, it added a new file and deleted it from the old location.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/_tests_/ecert-full-time-process-integration.scheduler.e2e-spec.ts (hunk 7 above)
Comment (ann-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/_tests_/ecert-full-time-process-integration.scheduler.e2e-spec.ts:84]: resolving since, its not scope of the PR

File: sources/packages/backend/apps/queue-consumers/test/helpers/testing-modules/testing-modules-helper.ts [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+/**
+ * Created the Queue Consumers root module needed to perform the E2E tests.
+ * @returns creation results with objects to support E2E tests.
+ */
+export async function createTestingAppModule(): Promise<CreateTestingModuleResult> {
+  const sshClientMock = createMock<Client>();
+  const module: TestingModule = await Test.createTestingModule({
+    imports: [QueueConsumersModule, createZeebeModuleMock()],
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/queue-consumers/test/helpers/testing-modules/testing-modules-helper.ts:30]: Will the queueModule in QueueConsumersModule look for redis?

File: sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+    beforeEach(() => {
+      jest.resetAllMocks();
+    });
+
+    it("Should cancel the assessment pending disbursements and rollback overawards when the cancelled application has overawards and also one sent and one pending disbursements.", async () => {
+      // Arrange
+      const workflowInstanceId = faker.random.number({
+        min: 1000000000,
+        max: 9999999999,
+      });
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts:61]: we can put the .toString() here to have the workflowInstanceId as string to use in other places.

```ts

 const workflowInstanceId = faker.random.number({
        min: 1000000000,
        max: 9999999999,
      }).toString();
```

File: sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+      expect(job.discard).toBeCalled();
+    });
+
+    it("Should throw an error and call job.discard when the assessment id was not found.", async () => {
+      // Arrange
+      const assessmentId = 9999;
+      const errorMessage = `Assessment id ${assessmentId} was not found.`;
+      const error = new Error(errorMessage);
+      // Queued job.
+      const job = createMock<Job<CancelAssessmentQueueInDTO>>({
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts:180]: May or may not be part of the PR, we can have a centralized method to get the jobMock which can have job input type, data and anything else which is required.

File: sources/packages/backend/apps/queue-consumers/test/helpers/testing-modules/testing-modules-helper.ts (hunk 10 above)
Comment (dheepak-aot) [reply sources/packages/backend/apps/queue-consumers/test/helpers/testing-modules/testing-modules-helper.ts:30]: I saw that the tests are using redis. Is there a way to tests the Processor class alone without involving redis? 
image

Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/test/helpers/testing-modules/testing-modules-helper.ts:30]: The current test files on this PR are not dependent on Redis.
I believe that we need to bring the `MockedQueueModule` from the API into lib and use it for both.

File: sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+        {
+          applicationStatus: ApplicationStatus.Completed,
+          createSecondDisbursement: true,
+        },
+      );
+      // Adjust assessment.
+      application.applicationStatus = ApplicationStatus.Cancelled;
+      const studentAssessment = application.currentAssessment;
+      studentAssessment.assessmentWorkflowId = workflowInstanceId.toString();
+      await studentAssessmentRepo.save(application.currentAssessment);
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts:75]: Just to check, is the application status updated through the cascade by saving the assessment?

I was assuming that to the cascade to happen, we muse do like this. `studentAssessment .application = application`

File: sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+      });
+
+      // Act
+      const result = await processor.cancelAssessment(job);
+
+      // Assert
+      expect(zbClientMock.cancelProcessInstance).toBeCalledWith(
+        workflowInstanceId.toString(),
+      );
+      expect(result.summary).toContain(
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts:100]: We can expect the whole object as we do in other e2e tests like this if the order of summary values are guaranteed.

image

File: sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts (hunk 13 above)
Comment (guru-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts:75]: Is this because the cascade happens only for update? as per model?

**Comment by dheepak-aot** [comment]
Thanks for creating E2E test foundation for queue consumers. Added some comments.

File: sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts (hunk 13 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts:75]: Line 72 was added by mistake during a conversation with @ann-aot. I will be removing it.

Comment (dheepak-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts:75]: If that is the case, I am confused here. If application status in not either cancelled or overwritten, how come the processor will not throw error as per this logic? 

image

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/_tests_/ecert-full-time-process-integration.scheduler.e2e-spec.ts [hunk 15]
```diff
@@ -0,0 +1,263 @@
...
+        student: savedStudent,
+      });
+      await studentRepo.save(savedStudent);
+      // MSFAA Number.
+      const savedMSFAANumber = await msfaaNumberRepo.save(
+        createFakeMSFAANumber({ student: savedStudent }),
+      );
+      // Create and save application.
+      const fakeApplication = createFakeApplication({ student: savedStudent });
+      fakeApplication.applicationNumber = "ECERT_TEST";
```
Comment (andrepestana-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/_tests_/ecert-full-time-process-integration.scheduler.e2e-spec.ts:95]: Why is it important to set the applicationNumber here and not use the random number?

**Review by andrepestana-aot — COMMENTED** [review]
Great job. I just added a question.

File: sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts (hunk 13 above)
Comment (ann-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts:75]: @dheepak-aot application status is `canceled`, that was leftover from our conversation.

File: sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts (hunk 14 above)
Comment (ann-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts:100]: I do not have a hard feeling IMO, we are matching what we want for the test in the PR.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/_tests_/ecert-full-time-process-integration.scheduler.e2e-spec.ts (hunk 15 above)
Comment (dheepak-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/_tests_/ecert-full-time-process-integration.scheduler.e2e-spec.ts:95]: @andrepestana-aot as per PR description, e-cert is not in the scope of PR.

File: sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts (hunk 13 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts:76]: Code adjusted.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/_tests_/ecert-full-time-process-integration.scheduler.e2e-spec.ts (hunk 15 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/_tests_/ecert-full-time-process-integration.scheduler.e2e-spec.ts]: I removed the line 95 as it was not important for the test and to avoid confusion in the future.

File: sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts (hunk 11 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts]: Changed.

File: sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts (hunk 12 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts:184]: I started creating it but I reverted it. The `createMock` is already pretty straight forward to be used and the options offered make it pretty easy to use also. If we feel a need in the future we may create a helper, IMO.

File: sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts (hunk 14 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts]: I made the change.
I was unwilling to validate the complete output for the queue consumers because, different from the API E2E where the return data is the most essential part, I considered the processing and DB as the most important and the output log as secondary information.

File: sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts (hunk 2 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/assessment/_tests_/cancel-application-assessment.processor.e2e-spec.ts:69]: Glad that we have the conversation. As per the chat, we will keep this one as it is.

File: sources/packages/backend/apps/queue-consumers/test/helpers/testing-modules/testing-modules-helper.ts (hunk 10 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/test/helpers/testing-modules/testing-modules-helper.ts]: As we discussed, Redis was there due to the BullModule connection created during module creation time. As discussed in the Team meeting, a workaround was created to allow the QueueModule to be entirely mocked and replaced.

**Review by andrepestana-aot — APPROVED** [review]

File: sources/packages/backend/apps/queue-consumers/test/helpers/testing-modules/testing-modules-helper.ts (hunk 10 above)
Comment (dheepak-aot) [reply sources/packages/backend/apps/queue-consumers/test/helpers/testing-modules/testing-modules-helper.ts]: Thanks for doing the work to remove redis.

**Review by ann-aot — COMMENTED** [review]
👍

**Review by dheepak-aot — APPROVED** [review]
Thanks for making the changes and walking through the replace module code. LGTM 👍

**Review by guru-aot — APPROVED** [review]
LGTM, nice work @andrewsignori-aot
